### PR TITLE
participant-integration-api: Move `TrackerMap` code around. [KVL-1009]

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
This is pure refactoring to:

- make `TrackerMap` a subtype of `Tracker`, by passing the tracker creation code into the constructor,
- make `TrackerMap` generic over the key, because it doesn't need to know the contents,
- move `Tracker.WithLastSubmission` to `TrackerMap.TrackerWithLastSubmission`, because it's only used by `TrackerMap`,
- move the `TrackerMap` scheduler into `TrackerMap.selfCleaning`,
- move `TrackerMap` construction into `ApiCommandService.create`, and
- move a bunch of other code around for convenience.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala [hunk 1]
```diff
@@ -66,7 +67,8 @@ private[apiserver] final class ApiCommandService private (
 
   private val logger = ContextualizedLogger.get(this.getClass)
 
-  private val submissionTracker: TrackerMap = TrackerMap(configuration.retentionPeriod)
+  private val submissionTracker: TrackerMap[TrackerKey] =
+    new TrackerMap(configuration.retentionPeriod, getTrackerKey, newTracker)
```
Comment (nicu-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala:71]: any reason to not receive this through the constructor? we would be able to extract `newTracker` as well then

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala [hunk 2]
```diff
@@ -203,13 +203,16 @@ object CommandsValidator {
...
-      else
-        commands.actAs.toSet + commands.party
-    val effectiveReadAs = commands.readAs.toSet -- effectiveActAs
-    Submitters(effectiveActAs, effectiveReadAs)
+    val actAs = effectiveActAs(commands)
+    val readAs = commands.readAs.toSet -- actAs
+    Submitters(actAs, readAs)
+  }
+
+  def effectiveActAs(commands: ProtoCommands): Set[String] = {
```
Comment (hubert-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:211]: We can get rid of these braces. Someone may also want to add some for the `if` expression ;)

**Review by hubert-da — APPROVED** [review]
LGTM

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala:71]: The `submissionTracker`? Works for me.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:211]: I like to skip the braces for simple `if` expressions.

Comment (hubert-da) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:211]: me 2 :)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerMap.scala [hunk 3]
```diff
@@ -59,17 +63,23 @@ private[services] final class TrackerMap(retentionPeriod: FiniteDuration)(implic
     }
   }
 
-  def track(submitter: TrackerMap.Key, request: SubmitAndWaitRequest)(
-      newTracker: => Future[Tracker]
-  )(implicit ec: ExecutionContext): Future[Either[TrackedCompletionFailure, CompletionSuccess]] =
+  def track(
```
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerMap.scala:65]: `override`

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandService.scala]: I pulled it out, but moved the scheduled task inwards, into `TrackerMap.selfCleaning`.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala [hunk 4]
```diff
@@ -203,13 +203,16 @@ object CommandsValidator {
   }
 
   def effectiveSubmitters(commands: ProtoCommands): Submitters[String] = {
```
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:205]: Yeah, I'd prefer if we didn't, but :shrug:

**Review by fabiotudone-da — COMMENTED** [review]
Same consideration about testing as in #10651.

**Comment by SamirTalwar-DA** [comment]
I had added tests in a branch which I was planning on opening as a PR later. (A lot of these changes were intended to get `TrackerMap` into a testable state.) However, I think you're right; they belong as part of these changes. So I have pulled the commit that adds existing test coverage from that branch.

Later commits from that branch add coverage and fix a few edge cases, which I think I'll keep separate.

**Comment by SamirTalwar-DA** [comment]
Gonna merge this in the interest of getting the next couple PRs in but please feel free to comment further.
